presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java

```diff
@@ -0,0 +1,382 @@
+        }
+        SessionPropertyMetadata<?> metadata = sessionProperty.getMetadata();
+        if (metadata.getJavaType() != type) {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, String.format("Property %s is type %s, but requested type was %s", name,
+                    metadata.getJavaType().getName(),
+                    type.getName()));
+        }
+        if (value == null) {
+            return type.cast(metadata.getDefaultValue());
+        }
+        Object objectValue = deserializeSessionProperty(metadata.getSqlType(), value);
+        try {
+            return type.cast(metadata.decode(objectValue));
+        }
+        catch (PrestoException e) {
+            throw e;
+        }
+        catch (Exception e) {
+            // the system property decoder can throw any exception
+            throw new PrestoException(INVALID_SESSION_PROPERTY, name + " is invalid: " + value, e);
+        }
+    }
+
+    @NotNull
+    public static String serializeSessionProperty(Type type, Object value)
+    {
+        if (value == null) {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "Session property can not be null");
+        }
+        if (BooleanType.BOOLEAN.equals(type)) {
+            return value.toString();
+        }
+        if (BigintType.BIGINT.equals(type)) {
+            return value.toString();
+        }
+        if (DoubleType.DOUBLE.equals(type)) {
+            return value.toString();
+        }
+        if (VarcharType.VARCHAR.equals(type)) {
+            return value.toString();
+        }
+        if (type instanceof ArrayType || type instanceof MapType) {
+            return getJsonCodecForType(type).toJson(value);
+        }
+        throw new PrestoException(INVALID_SESSION_PROPERTY, String.format("Session property type %s is not supported", type));
+    }
+
+    @NotNull
+    private static Object deserializeSessionProperty(Type type, String value)
+    {
+        if (value == null) {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, String.format("Session property can not be null"));
+        }
+        if (VarcharType.VARCHAR.equals(type)) {
+            return value;
+        }
+        if (BooleanType.BOOLEAN.equals(type)) {
+            return Boolean.valueOf(value);
+        }
+        if (BigintType.BIGINT.equals(type)) {
+            return Long.parseLong(value);
+        }
+        if (DoubleType.DOUBLE.equals(type)) {
+            return Double.parseDouble(value);
+        }
+        if (type instanceof ArrayType || type instanceof MapType) {
+            return getJsonCodecForType(type).fromJson(value);
+        }
+        throw new PrestoException(INVALID_SESSION_PROPERTY, String.format("Session property type %s is not supported", type));
+    }
+
+    // there is no clean way to do this
+    @SuppressWarnings("rawtypes")
+    private static JsonCodec getJsonCodecForType(Type type)
```

Comment:
There is no clean way to do this. `?` gives an error at the use since `Object` is not `?`.  `Object` gives an error at codec creation since `String` is not `Object`.  The `extends` and `super` versions don't with either.